keras_cv/models/segmentation/fcn.py

```diff
@@ -0,0 +1,268 @@
+                self.dense_convs.append(
+                    keras.layers.Conv2D(
+                        filters=self.num_classes,
+                        strides=(1, 1),
+                        activation="relu",
+                        padding="same",
+                    )
+                )
+            self.output_layer = keras.layers.UpSampling2D(
+                size=(target_height_factor, target_width_factor),
+                data_format="channels_last",
+                interpolation="bilinear",
+            )
+
+    def call(self, x):
+        if self.model_architecture == "fcn8s":
+            pool3_output = self.pool3(x)
+            pool4_output = self.pool4(x)
+            pool5_output = self.pool5(x)
+
+            for layer in self.dense_convs:
+                pool5_output = layer(pool5_output)
+            pool5_output = keras.layers.UpSampling2D(
+                size=(2, 2), data_format="channels_last", interpolation="bilinear"
+            )(pool5_output)
+
+            intermediate_pool_output = keras.layers.Add()[pool4_output, pool5_output]
+            intermediate_pool_output = keras.layers.UpSampling2D(
+                size=(2, 2), data_format="channels_last", interpolation="bilinear"
+            )(intermediate_pool_output)
+
+            final_pool_output = keras.layers.Add()[
+                pool3_output, intermediate_pool_output
+            ]
+
+            output_layer = keras.layers.UpSampling2D(
+                size=(8, 8), data_format="channels_last", interpolation="bilinear"
+            )
+            return output_layer(final_pool_output)
+
+        elif self.model_architecture == "fcn16s":
+            pool4_output = self.pool4(x)
+            pool5_output = self.pool5(x)
+
+            for layer in self.dense_convs:
+                pool5_output = layer(pool5_output)
+            pool5_output = keras.layers.UpSampling2D(
+                size=(2, 2), data_format="channels_last", interpolation="bilinear"
+            )(pool5_output)
+
+            final_pool_output = keras.layers.Add()[
+                pool4_output, intermediate_pool_output
+            ]
+
+            output_layer = keras.layers.UpSampling2D(
+                size=(16, 16), data_format="channels_last", interpolation="bilinear"
+            )
+            return output_layer(final_pool_output)
+
+        elif self.model_architecture == "fcn32s":
+            pool5_output = self.pool5(x)
+
+            for layer in self.dense_convs:
+                pool5_output = layer(pool5_output)
+            output_layer = keras.layers.UpSampling2D(
+                size=(32, 32), data_format="channels_last", interpolation="bilinear"
+            )
+            return output_layer(pool5_output)
+
+        else:
+            x = self.backbone(x)
+            for layer in self.dense_convs:
+                x = layer(x)
+            return self.output_layer(x)
+
+    def train_step(self, data):
```

Comment:
Why use a custom training loop?

Reply:
Well, not sure about how `model.fit()` would otherwise handle the training. Also, in the `deeplab.py` file too, the training has been delegated to a custom loop. Is there a way to handle this without making a custom loop?

Reply:
DeepLab actually doesn't need the custom training loop. #1119 has a run without a custom training loop as part of the refactor. 

I'd still wait on feedback for that specific one since the loop contains a couple of non-standard steps. Here, the loop seems to just run (pseudocode):

```
y_pred = model(inputs)
loss = criterion(y_pred, y_true)
optimizer.minimize(loss)
```

Which is pretty standard. The default `model.fit()` should cover it fully, I think.

Reply:
That makes sense, but I guess we can wait for some feedback from keras-team on this.

Reply:
@tanzhenyu Thoughts?

Reply:
The reason why deeplab needs an overriding train_step is due to weight regularization. In the latest tf release (2.10 and 2.11), the new experimental SGD accepts `weight_decay` as parameter, so ideally we don't need an overriding train_step.
However, I have not fully tested the new versions yet (I think there's some behavior misgap for SGD between 2.10 and 2.11 as well)

Reply:
That being said - since this loop just does the basic operations - try training with and without it. `model.fit()` should cover this one.
It it needs weight regularization - keeping it (or using `weight_decay` if that can work) sounds better.

If we can avoid a custom loop, IMO, we should to make the code simpler and approachable to future users.